.ci/azure-pipelines/azure-pipelines.yaml

```diff
@@ -1,3 +1,19 @@
+trigger:
+  paths:
+    exclude:
+    - doc
```

Comment:
Adding this will result in the documentation stage not running.

The aim is to have the documentation CI run regardless, but the code CI (Ubuntu, Mac and Windows) to run only when code is changed (and doc is not code that's compiled in them)


I think properly implementing the feature will require one of the following
* adding a separate pipeline for documentation
* separate stage in other CI to use git to check which files are modified and skip if so

Reply:
I'm sorry to cause some trouble, and increase your workload of review. I made the commit and assumed that any updates to doc(contains doc directory, README.md and etc) won't cause CI(includes documentation CI) to run at once. Whenever there is a commit with code files changed, the main CI(build and documentation) will run and the documentation repo will be updated by the way. 

Maybe that's not a nice way. And I think "adding a separate pipeline for documentation" will be a better choice and separating the main pipeline into pieces with clearer purposes will bring more convenience to CI. 

I'll do it soon. Thanks for your advice!

Reply:
For separate CI, I don't know if it's possible to run a pipeline after another, but the following constraint would be nice to have:

The tutorials stage should be run after the documentation stage, only if the ubuntu build CI goes green.

Reply:
I follow the rules you commented.

In the latest commit, I separate both **documentation stage** and **tutorials stage** from **main build pipeline** to **documentation pipeline**. And I introduce [pipeline trigger](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/pipeline-triggers?view=azure-devops&tabs=yaml), which allows running a pipeline after another, to the new pipeline to ensure the tutorials stage will be triggered whenever the "Build GCC on Ubuntu" stage are completed successfully.

The expected behavior is listed below:
- code changes: **main build CI** run, and trigger the complete documentation CI only after "Build GCC" is OK.
- doc changes: only trigger documentation stage, and tutorials stage won't run. (which is current behavior of CI)

I've tested all those behaviors in [another branch](https://github.com/ueqri/pcl/commits/test-feature-on-minimal-CI) using simplified pipelines but the same structure.

And the result of the running is here: https://dev.azure.com/ueqri-ci/pcl-test/_build?view=runs

Reply:
Thanks @ueqri 

Just a minor change:
* Tutorial stage should run in the end: regardless of Documentation pipeline or Build pipeline (since it serves 2 purposes: making sure tutorials work, with the latest code)

Apologies for not making it clear beforehand.

Situations:
1. Doc changes: format -> generate doc -> run tutorials stage
2. Code changes: format-> (run build, generate doc) -> run tutorials stage

For reference, currently, our pipeline performs (format -> build -> tutorials and doc in parallel) for every change.

I hope this clarifies the situation a bit.

Reply:
I think I got it, the tutorials stage should always run if everything works well(especially "build gcc on ubuntu").

BTW, I guess the "formatting stage" is unnecessary for docs pipeline, could I just run docs pipeline without it?

Reply:
I also find that the artifact(named "formatting") published by "formatting stage" has not been used in other files or pipelines. Is it an issue to be addressed or a feature as preparation for future work? Could I do something to help on this?

Reply:
Ah. The thing is, we have not applied formatting to docs yet. The formatting is applied slowly across different modules (see the `.dev/format.sh` for details). Instead of modifying the pipeline later, we can have the required pipeline before we have formatting on docs.

>  Is it an issue to be addressed or a feature as preparation for future work? Could I do something to help on this?

It's just a patch that could be applied (useful, if we could somehow showcase it). I wanted a bot to auto-comment the diff, using the suggestion feature of GitHub, but it turned out that I don't have time for that 😆 That's something for a different PR though

Reply:
Thanks! I understand the structure of current CI better! 

I'll make the change for the following situations:
> 1. Doc changes: format -> generate doc -> run tutorials stage
> 2. Code changes: format-> (run build, generate doc) -> run tutorials stage

BTW, I'm interested in the bot to auto-comment the diff :laughing: and I want to learn and test it on my branch first.